py/surveyqa/summary.py

```diff
@@ -236,17 +232,60 @@ def last(arr):
         return arr[len(arr)-1]
     return exposures.group_by("TILEID").groups.aggregate(last)
 
-tzone = TimezoneInfo(utc_offset = -7*u.hour)
+utc_offset = -7*u.hour
+tzone = TimezoneInfo(utc_offset = utc_offset)
 t1 = Time(58821, format='mjd', scale='utc')
 t = t1.to_datetime(timezone=tzone)
 
-def get_surveyprogress(exposures, tiles, line_source, hover_follow, width=250, height=250, min_border_left=50, min_border_right=50):
+def get_progress(exposures, tiles, program):
+    '''Get survey and tile progress for a given program
+
+    Args:
+        exposures: Table with columns PROGRAM, TILEID, MJD
+        tiles: Table with columns TILEID, EXPOSEFAC
+        program: str program name to filter
+
+    Returns (time, survey_progress, tile_progress)
+
+    time = array of datetime objects;
+    survey_progress = array of EXPOSEFAC-weighted progress on observing tiles;
+    tile_progress = array of number of tiles observed
+    '''
+
+    keep = exposures['PROGRAM'] == program
+    exposures = exposures['TILEID', 'MJD'][keep]
+    keep = tiles['PROGRAM'] == program
+    tiles = tiles['TILEID', 'EXPOSEFAC'][keep]
+
+    #- Create table of last exposure for each tile, sorted by MJD
+    finished_tiles = exposures.group_by('TILEID').groups.aggregate(np.max)
+    finished_tiles = join(finished_tiles, tiles, keys='TILEID', join_type='left')
+    finished_tiles.sort('MJD')
+
+    #- Convert MJD to list of datetimes for bokeh
+    t1 = Time(finished_tiles['MJD'], format='mjd', scale='utc')
+    # t = t1.to_datetime(timezone=tzone)
+    # local non-timezone aware time, instead of much slower timezone aware
+    t = (t1 + utc_offset).to_datetime()
+
+    #- Tile progress is just counting tiles
+    tile_progress = np.arange(len(t))
+
+    #- Survey progress is weighted by EXPOSEFAC
+    survey_progress = np.cumsum(finished_tiles['EXPOSEFAC']) / np.sum(tiles['EXPOSEFAC'])
+
+    return t, tile_progress, survey_progress
+
+
+def get_surveyprogress_plot(progress, line_source, hover_follow, width=250, height=250, min_border_left=50, min_border_right=50):
     '''
     Generates a plot of survey progress (EXPOSEFAC) vs. time
 
+    TODO: update docs for what "progress" is
+
     Args:
-        exposures: Table of exposures with columns "PROGRAM", "TILEID", "MJD"
-        tiles: Table of tile locations with columns "TILEID", "EXPOSEFAC", "PROGRAM"
+        progress: dict keyed by DARK, GRAY, BRIGHT, with values
+            (time, survey_progress, tile_progress) from get_progress
```

Comment:
Optimization made the interface less clean: instead of directly passing in `exposures` and `tiles`, now we have to pre-calculate `progress = get_progress(exposures, tiles, program)` and pass that in instead.  The benefit is that `progress` gets re-used by another function too without having to repeat much of the calculation.

This is a good example of waiting until the code algorithmically works before doing much optimization, since this optimization made it somewhat less clean and flexible.